smoothscroll.js

```diff
@@ -1,156 +1,237 @@
         startX = startY = endX = endY = undefined;
         return;
       }
 
-      frame = requestAnimationFrame(step);
-    };
-
-    if (frame) cancelAnimationFrame(frame);
-    frame = requestAnimationFrame(step);
-  }
-
-  window.scroll = window.scrollTo = function(x, y, scrollOptions) {
-    if (shouldBailOut(scrollOptions))
-      return originalScrollTo(x, y);
-    return smoothScroll(x, y);
-  };
-
-  window.scrollBy = function(x, y, scrollOptions) {
-    if (shouldBailOut(scrollOptions))
-      return originalScrollBy(x, y);
-
-    var sx = window.pageXOffset;
-    var sy = window.pageYOffset;
-
-    return smoothScroll(x + sx, y + sy);
-  };
-
-  var elementRects, scrollableParent;
-  function scrollElement(el, x, y) {
-    el.scrollTop = y;
-    el.scrollLeft = x;
+      frame = w.requestAnimationFrame(step);
+    }
   }
 
-  function scroll(el, endCoords) {
-    var sx = el.scrollLeft;
-    var sy = el.scrollTop;
-
-    var x = endCoords.left;
-    var y = endCoords.top;
-
-    if (typeof startX === 'undefined') {
+  /*
+   * scrolls inside an element with a smooth behavior
+   * @method smoothScrollElement
+   * @params {Node} el
+   * @params {Object} endCoords
+   */
+  function smoothScrollElement(el, endCoords) {
+    var sx = el.scrollLeft,
+        sy = el.scrollTop,
+        x = endCoords.left,
+        y = endCoords.top,
+        startTime = now();
+
+    if (startX === undefined) {
       startX = sx;
       startY = sy;
       endX = endCoords.left;
       endY = endCoords.top;
     }
 
-    var startTime = now();
+    // cancel frame is there is an scroll event happening
+    if (frame) {
+      w.cancelAnimationFrame(frame);
+    }
 
-    var step = function() {
-      var time = now();
-      var elapsed = (time - startTime) / SCROLL_TIME;
+    frame = w.requestAnimationFrame(step);
+
+    // scroll looping over a frame
+    function step() {
+      var time = now(), value, cx, cy,
+          elapsed = (time - startTime) / SCROLL_TIME;
+
+      // avoid elapsed times higher than one
       elapsed = elapsed > 1 ? 1 : elapsed;
 
-      var value = ease(elapsed);
-      var cx = sx + ( x - sx ) * value;
-      var cy = sy + ( y - sy ) * value;
+      value = ease(elapsed);
+      cx = sx + ( x - sx ) * value;
+      cy = sy + ( y - sy ) * value;
 
       scrollElement(el, cx, cy);
 
+      // return if end points have been reached
       if (cx === endX && cy === endY) {
         startX = startY = endX = endY = undefined;
         return;
       }
 
-      frame = requestAnimationFrame(step);
-    };
-
-    if (frame) cancelAnimationFrame(frame);
-    frame = requestAnimationFrame(step);
+      frame = w.requestAnimationFrame(step);
+    }
   }
 
-  function findScrollableParent(el) {
-    if (el.clientHeight < el.scrollHeight ||
-        el.clientWidth < el.scrollWidth)
-      return el;
-    return findScrollableParent(el.parentNode);
-  }
+  // ORIGINAL METHODS OVERRIDES
+  // window.scroll and window.scrollTo
```

Comment:
- ```arguments[0]``` can be object or number, if it's not obejct default behavior will be called
- ```arguments[0]```can be an object but have auto or undefined behavior so we can send left and top coordenates or just the first two arguments of ```window.scroll``` to support both situations in the same condition